src/relay/backend/contrib/vsi_npu/codegen.cc

```diff
@@ -0,0 +1,425 @@
+      pre_opsetup->specs_[0].datatype_ == tvx::DataType::BOOL8) {
+    return;
+  }
+
+  tvx::Quantization& now_quant_info = now_opsetup->specs_[0].quantization_;
+
+  std::vector<int32_t> zps;
+  std::vector<float> scales;
+  if (now_quant_info.Type() == tvx::QuantType::NONE) {
+    zps = {0};
+    scales = {1.0};
+    now_quant_info.SetType(tvx::QuantType::ASYMMETRIC).SetScales({1.0}).SetZeroPoints({0});
+  } else {
+    zps = now_quant_info.ZeroPoints();
+    scales = now_quant_info.Scales();
+  }
+
+  while (ptr_callback &&
+         op_tb[ptr_callback->expr_]->specs_[0].quantization_.ZeroPoints().size() == 0) {
+    Expr expr = ptr_callback->expr_;
+    auto datatype = GetTvxType(expr->checked_type().as<TensorTypeNode>()->dtype);
+    if (datatype != tim::vx::DataType::INT32) {
+      op_tb[expr]
+          ->specs_[0]
+          .quantization_.SetType(tvx::QuantType::ASYMMETRIC)
+          .SetScales(scales)
+          .SetZeroPoints(zps);
+    }
+    ptr_callback = ptr_callback->ptr_pre_callback_;
+  }
+}
+
+template <typename T, typename T2>
+void attribute_transform(const T &attrs, T2 &attrs_num) {
+
+  std::transform(attrs.begin(), attrs.end(), attrs_num.begin(),
+                 [](const PrimExpr &attrs_num) {
+                   return static_cast<uint32_t>(
+                       attrs_num.as<IntImmNode>()->value);
+                 });
+};
+
+std::shared_ptr<tvx::Tensor> createVxOPerand(TensorInfoTable tensor_info,
+                                             Expr expr, tvx::Graph *graph,
+                                             uint32_t idx = 0) {
+  auto tensor_spec = tensor_info[expr][idx];
+  void *data = expr->IsInstance<ConstantNode>()
+                   ? expr.as<ConstantNode>()->data->data
+                   : nullptr;
+  return data == nullptr ? graph->CreateTensor(tensor_spec)
+                         : graph->CreateTensor(tensor_spec, data);
+};
+
+static std::vector<tim::vx::TensorSpec>
+GetTimVxTensorSpec(const TupleTypeNode *tuple) {
+  auto input_node_tensors = tuple->fields;
+
+  std::vector<tim::vx::TensorSpec> specs;
+  uint32_t input_node_num = input_node_tensors.size();
+  for (uint32_t i = 0; i < input_node_num; i++) {
+    std::cout << "GetTimVxTensorSpec: " << input_node_tensors[i].as<TensorTypeNode>() << std::endl;
+    tim::vx::ShapeType shape;
+    std::transform(input_node_tensors[i].as<TensorTypeNode>()->shape.rbegin(),
+                   input_node_tensors[i].as<TensorTypeNode>()->shape.rend(),
+                   std::back_inserter(shape), [](const PrimExpr &dim) {
+                     return static_cast<int>(dim.as<IntImmNode>()->value);
+                   });
+
+    auto dtype = input_node_tensors[i].as<TensorTypeNode>()->dtype;
+    auto dataType = GetTvxType(dtype);
+
+    tim::vx::TensorSpec spec(dataType, shape,
+                             tim::vx::TensorAttribute::OUTPUT);
+    specs.push_back(spec);
+  }
+  return specs;
+}
+
+using namespace backend;
```

Comment:
prefer to avoid this https://google.github.io/styleguide/cppguide.html#Namespaces

Reply:
Will fix.